### PR TITLE
[#4766] Win32: Added support NUMA configuration with multiple count CPU sockets

### PR DESCRIPTION
**Pull request by GermanAizek**
Fixed this issue: https://github.com/git-for-windows/git/issues/4766

### CODE REVIEW

File: thread-utils.c [hunk 1]
```diff
@@ -18,6 +18,67 @@
 #  endif
 #endif
 
+#ifdef GIT_WINDOWS_NATIVE
+/// defines the GetLogicalProcessorInformationEx function
+typedef BOOL (*WINAPI glpie_t)(
+	LOGICAL_PROCESSOR_RELATIONSHIP,
+	PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX,
+	PDWORD);
```
Comment (dscho) [inline thread-utils.c:26]: Please move this code to `compat/win32/` and use `lazyload.h` (imitate existing code that [declares](https://github.com/git-for-windows/git/blob/ad0bbfffa543db6979717be96df630d3e5741331/compat/mingw.c#L2807-L2808) and [initializes](https://github.com/git-for-windows/git/blob/ad0bbfffa543db6979717be96df630d3e5741331/compat/mingw.c#L2812) the function pointer).

File: thread-utils.c [hunk 2]
```diff
@@ -18,6 +18,67 @@
 #  endif
 #endif
 
+#ifdef GIT_WINDOWS_NATIVE
+/// defines the GetLogicalProcessorInformationEx function
```
Comment (dscho) [inline thread-utils.c:22]: Please avoid C++-style comments. The existing code does not use them.

File: thread-utils.c [hunk 3]
```diff
@@ -18,6 +18,67 @@
...
+typedef BOOL (*WINAPI glpie_t)(
+	LOGICAL_PROCESSOR_RELATIONSHIP,
+	PSYSTEM_LOGICAL_PROCESSOR_INFORMATION_EX,
+	PDWORD);
+
+// Classic way to get number of cores in windows.
+static size_t get_windows_compatible_n_cores() {
```
Comment (dscho) [inline thread-utils.c:29]: Let's not add an extra function here. Instead, let's declare a single `win32_online_cpus(void)` function in a header file in `compat/win32/` and implement it in a `.c` file. Or put them into `compat/mingw.h`/`compat/mingw.c`.

File: thread-utils.c [hunk 4]
```diff
@@ -18,6 +18,67 @@
...
+	size_t n_cores = 0;
+	DWORD length = 0;
+	HMODULE kernel32 = GetModuleHandleW(L"kernel32.dll");
+
+	glpie_t GetLogicalProcessorInformationEx = (glpie_t)GetProcAddress(kernel32, "GetLogicalProcessorInformationEx");
+	if (!GetLogicalProcessorInformationEx) {
+		return 0;
+	}
+
+	GetLogicalProcessorInformationEx(RelationAll, NULL, &length);
```
Comment (dscho) [inline thread-utils.c:54]: Let's not forget to check the return value.

**Review by dscho — CHANGES_REQUESTED** [review]
Thank you for contributing this PR! While NUMA is probably still rather uncommon, it'll be good to have the support (once the code compiles, that is).

**Comment by GermanAizek** [comment]
@dscho, I will not be able to fix this branch without testing, I have completely switched to Linux and FreeBSD. Is there anyone who can test this branch with NUMA nodes on any motherboard with two or more processors?

**Comment by GermanAizek** [comment]
@dscho When I did the fix in this PR, it accurately detected count threads correctly.

**Comment by GermanAizek** [comment]
@dscho which is very good on Linux there are no such problems and it has been originally laid down by POSIX library, Microsoft provides many methods in Winapi to get count threads, obviously software developers use the most common in tutorials and on stackoverflow, although this is wrong. Because of this on any Windows, only 50% CPU util is used with two NUMA nodes, 25% with 4 nodes, with 6 numa nodes 16.66%

**Comment by dscho** [comment]
> I will not be able to fix this branch without testing, I have completely switched to Linux and FreeBSD. Is there anyone who can test this branch with NUMA nodes on any motherboard with two or more processors?

That's unfortunate, I have no such system to test, either.

Let's just close this PR until the time when (if?) someone with such a system can test and work on it.

**Pull request closed without merging**